### PR TITLE
MAINT: Add `dev.py` for building and installing scipy with meson

### PR DESCRIPTION
**Pull request by Smit-create**
See: #36 

cc @rgommers

### CODE REVIEW

**Comment by Smit-create** [comment]
I have added a few options additionally here https://github.com/rgommers/scipy/pull/94/commits/a323c17563eb1e087138527b4224d7baf965799d like: Customising `installdir` name and whether to setup meson build at each execution.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -77,17 +77,9 @@ jobs:
...
-        meson setup build --prefix=$PWD/${{ env.INSTALLDIR }} --werror
-
-    - name: Build SciPy
-      run: |
-        ninja -C build -j 2
-
-    - name: Install SciPy
-      run: |
-        meson install -C build
+        python dev.py -rb --installdir ${{ env.INSTALLDIR }} -j 2 --build-only --werror
```
Comment (rgommers) [inline .github/workflows/linux.yml:82]: This does look more concise. One thing I liked about the original is that the build and install steps are separate, so the ~1500 lines of output for the install step can be ignored easily when looking at build issues. Maybe that's something to keep - have to think about it more.

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -102,6 +94,4 @@ jobs:
     - name: Test SciPy
       run: |
         export PYTHONPATH="$PWD/installdir/lib/python3.9/site-packages/"
-        pushd ${{ env.INSTALLDIR }}
-        python -c "import scipy; import sys; sys.exit(int(not scipy.test(parallel=2)))"
-        popd
+        python dev.py -n -j 2
```
Comment (rgommers) [inline .github/workflows/linux.yml:97]: This is definitely much nicer

**Review by rgommers — COMMENTED** [review]
Thanks @Smit-create! My main question at this point is what parts of this file are unchanged and therefore don't need review now and can in the end be removed in favor of importing them from `runtests.py`?

**Comment by rgommers** [comment]
> whether to setup meson build at each execution.

That will actually give an error if you try to re-run `meson setup` for a directory that's already configured. So that may need something more subtle, or just not try. Probably if the build dir exists, just assume it's configured already.

**Comment by Smit-create** [comment]
> Probably if the build dir exists, just assume it's configured already.

Initially, I did the same. Will revert back the `setup-build` and `installdir` options. I think its standard to keep `installdir`

**Comment by rgommers** [comment]
`installdir` configurable seems fine to me. Also, it's possible to have multiple build dirs, in which case one does need multiple install dirs. But that's fine to ignore for now, let's first get the basic functionality merged.

**Comment by Smit-create** [comment]
> My main question at this point is what parts of this file are unchanged and therefore don't need review now and can in the end be removed in favor of importing them from `runtests.py`

Only, `setup_build`, `install_project` and `build_project` functions were added/modified. We can't import all the functions from `runtests.py` since some of them has a callback to `build_project`

**Comment by rgommers** [comment]
This is the trick you need to import from `runtests.py`: https://github.com/scipy/scipy/blob/master/scipy/_build_utils/__init__.py#L21

**Comment by Smit-create** [comment]
@rgommers `importlib` seems to be confusing. It fails on Linux tests here,  runs on macOS buts fails due to sys.path changes. I made a few changes(the latest commit pushed here) and got it running on my Local Machine (Ubuntu 18.04), but failures still exit on CI here.


```
$ python dev.py -n -s io
Running tests for scipy version:1.8.0.dev0+Unknown.Unknown, installed at:/home/smit/Smitlunagariya/scipy/installdir/lib/python3.9/site-packages/scipy
============================= test session starts ==============================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /home/smit/Smitlunagariya/scipy, configfile: pytest.ini
plugins: cov-2.12.1, xdist-2.3.0, forked-1.3.0
collected 606 items                                                            

../../installdir/lib/python3.9/site-packages/scipy/io/_harwell_boeing/tests/test_fortran_format.py . [  0%]
..........                                                               [  1%]
../../installdir/lib/python3.9/site-packages/scipy/io/_harwell_boeing/tests/test_hb.py . [  1%]
.                                                                        [  2%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_byteordercodes.py . [  2%]
.                                                                        [  2%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_mio.py . [  2%]
..................................................................x      [ 13%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_mio5_utils.py . [ 13%]
.....                                                                    [ 14%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_mio_funcs.py . [ 14%]
                                                                         [ 14%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_mio_utils.py . [ 15%]
.                                                                        [ 15%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_miobase.py . [ 15%]
                                                                         [ 15%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_pathological.py . [ 15%]
.                                                                        [ 15%]
../../installdir/lib/python3.9/site-packages/scipy/io/_matlab/tests/test_streams.py . [ 15%]
..........                                                               [ 17%]
../../installdir/lib/python3.9/site-packages/scipy/io/arff/tests/test_arffread.py . [ 17%]
..........................                                               [ 21%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_fortran.py . [ 22%]
..........                                                               [ 23%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_idl.py . [ 23%]
.................................................................        [ 34%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_mmio.py . [ 34%]
........................................................................ [ 46%]
.                                                                        [ 46%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_netcdf.py . [ 47%]
........................                                                 [ 50%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_paths.py . [ 51%]
..........                                                               [ 52%]
../../installdir/lib/python3.9/site-packages/scipy/io/tests/test_wavfile.py . [ 52%]
........................................................................ [ 64%]
........................................................................ [ 76%]
........................................................................ [ 88%]
.....................................................................    [100%]

======================= 605 passed, 1 xfailed in 12.28s ========================

```

**Comment by Smit-create** [comment]
Linux job now passes. macOS has some issues(The same problem was on Linux which was fixed by `sys.path` changes made from `importlib` in https://github.com/rgommers/scipy/pull/94/commits/f2fb973aeed6d242fb3c13c056c4f9eca5a400ff). @rgommers Can you please help with macOS failures?

**Comment by rgommers** [comment]
So the issue it's running into on macOS here is:
```
from . cimport cython_blas as blas_pointers
^
------------------------------------------------------------

scipy/linalg/_decomp_update.pyx:64:0: relative cimport beyond main package is not allowed
```
which is due to gh-31, and upstream issue https://github.com/mesonbuild/meson/issues/8961.

If there was a further issue with this PR, it's not made visible in the current CI log. @Smit-create is that what you were seeing before?

**Comment by Smit-create** [comment]
>  @Smit-create is that what you were seeing before?

I didn't get this error before.

**Comment by rgommers** [comment]
Okay then I'll re-run it once more.

**Comment by rgommers** [comment]
Ah yes, now the build completed and it shows:
```
Traceback (most recent call last):
  File "/Users/runner/work/scipy/scipy/dev.py", line 589, in <module>
    main(argv=sys.argv[1:])
  File "/Users/runner/work/scipy/scipy/dev.py", line 309, in main
    __import__(PROJECT_MODULE)
ModuleNotFoundError: No module named 'scipy'
Error: Process completed with exit code 1.
```

**Comment by rgommers** [comment]
Build log change looks good, thanks.

Locally `python dev.py` works for me. I think the issue is that the command is split in `--build-only` and `-n` (= `--no-build`, and the test command in `macos.yml` has:
```
    if not args.no_build:                                            
        site_dir = build_project(args)          
        sys.path.insert(0, site_dir)                       
        os.environ['PYTHONPATH'] = \                          
            os.pathsep.join((site_dir, os.environ.get('PYTHONPATH', '')))
```
So if  `PYTHONPATH` doesn't match the hardcoded value in `macos.yml`, things will fail. Maybe add some debug prints in CI to check the value of `PYTHONPATH` and the actual install path?

**Comment by Smit-create** [comment]
My guess for this failures on MacOS are:
```
Final sys path ['/Users/runner/work/scipy/scipy/installdir/lib/python3.9/site-packages', '/usr/local/miniconda/envs/scipy-dev/lib/python310.zip', '/usr/local/miniconda/envs/scipy-dev/lib/python3.10', '/usr/local/miniconda/envs/scipy-dev/lib/python3.10/lib-dynload', '/usr/local/miniconda/envs/scipy-dev/lib/python3.10/site-packages', '/usr/local/miniconda/envs/scipy-dev/lib/python3.10/site-packages/gmpy2-2.1.0rc1-py3.10-macosx-10.9-x86_64.egg']
```

It's probably because, macOS seems to be using `python3.10` where as we have hardcoded `python3.9 `in `export PYTHONPATH="$PWD/installdir/lib/python3.9/site-packages/"`

**Comment by Smit-create** [comment]
@rgommers Woahh, Tests are running!!!

**Comment by rgommers** [comment]
Awesome, that passes modulo some of the usual failures still in numpy `1.21.4` due to:
```
E   RuntimeWarning: invalid value encountered in reciprocal
```
The functions in `dev.py` that are present all seem to have modifications that are needed with Meson; the one I'm not sure about is `run_mypy` - does that not work if imported from `runtests`?

File: dev.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+        raise RuntimeError(
+            "Mypy not found. Please install it by running "
+            "pip install -r mypy_requirements.txt from the repo root"
+        ) from e
+
+    site_dir = build_project(args)
+    config = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)),
+        "mypy.ini",
+    )
```
Comment (Smit-create) [inline dev.py:522]: Here, `run_mypy` calls `build_project`, so, I haven't imported it from `runtests.py`.

Comment (rgommers) [reply dev.py:522]: Ah, makes sense, thanks.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes! Thanks @Smit-create

**Comment by rgommers** [comment]
As a follow-up, could you
- update `MESON_BUILD.md`?
- remove `mesondev.sh`
- consider whether the `export OMP_NUM_THREADS=2` bit in `mesondev.sh` needs to be added in the GitHub Actions workflows? It should improve test runtime in CI.

**Comment by Smit-create** [comment]
Yes, will do

**Pull request merged**